a2c/a2c.py

```diff
@@ -0,0 +1,169 @@
+import gym
+import numpy as np
+import torch
+import torch.nn as nn
+from torch.autograd import Variable
+
+CART_POLE_IN_NUM = 4
+CART_POLE_OUT_NUM = 2
+
+# ONLY NETWORK HERE  ==> IT'LL BE REPLACED WITH OUR NETWORK
+
+
+class Net(nn.Module):
+    def __init__(self, in_num, out_num):
+        super(Net, self).__init__()
+        self.policy = nn.Sequential(
+            nn.Linear(in_num, 50),
+            nn.ReLU(),
+            nn.Linear(50, 20),
+            nn.ReLU(),
+            nn.Linear(20, out_num),
+            nn.Softmax(dim=1))
+
+        self.value = nn.Sequential(
+            nn.Linear(in_num, 50),
+            nn.ReLU(),
+            nn.Linear(50, 20),
+            nn.ReLU(),
+            nn.Linear(20, 1))
+
+    def forward(self, x):
+        x = Variable(torch.from_numpy(x).float().unsqueeze(0))
+        action_prob = self.policy(x.float())
+        value = self.value(x.float())
+        return action_prob, value
+# NETWORK PART END
+
+
+class A2CTrainer:
+    def __init__(self, net, out_num, environment_name, batch_size,
+                 gamma, beta_entropy, learning_rate, clip_size):
+
+        self.net = net
+        self.out_num = out_num
+        self.env = gym.make(environment_name)
+        self.batch_size = batch_size
+        self.gamma = gamma
+        self.beta_entropy = beta_entropy
+        self.learning_rate = learning_rate
+        self.clip_size = clip_size
+        self.optimizer = torch.optim.Adam(
+            self.net.parameters(), lr=self.learning_rate)
+        self.estim_values = []
+        self.action_logarithms = []
+        self.rewards = []
+        self.entropy_loss = 0
+        self.Qvals = None
+        self.render = False
+
+    def choose_action(self, probabilities):
+        return np.random.choice(self.out_num, p=np.squeeze(probabilities))
+
+    def calculate_entropy(self, action_chance):
+        return -np.sum(np.mean(action_chance) * np.log(action_chance))
+
+    def calculate_logarithm_probs(self, probabilities, chosen_action):
+        return torch.log(probabilities.squeeze(axis=0)[chosen_action])
+
+    def clear_previous_batch_data(self):
+        self.estim_values = []
+        self.action_logarithms = []
+        self.rewards = []
+        self.entropy_loss = 0
+
+    def save_batch_data(self, reward, estim_value, logarithm_probs, entropy):
+        self.rewards.append(reward)
+        self.estim_values.append(estim_value)
+        self.action_logarithms.append(logarithm_probs)
+        self.entropy_loss += entropy
+
+    def calculate_qvals(self, Qval):
+        self.Qvals = np.zeros_like(self.estim_values)
+        for it in reversed(range(len(self.rewards))):
+            Qval = self.rewards[it] + self.gamma * Qval
+            self.Qvals[it] = Qval
+
+    def calculate_actor_loss(self, advantage):
+        return (-self.action_logarithms * advantage).mean()
```

Comment:
Wydaje mi się, że action_logarithms to lista i nie wiem dlaczego nie pluje się o ten minus. Proponuję owrappować ją w np.array lub zastosować od początku np.array zamiast listy